Python/thread_nt.h

```diff
@@ -266,6 +266,112 @@ PyThread_exit_thread(void)
+    if (key == TLS_OUT_OF_INDEXES) return FALSE;
+    *(DWORD*)parameter = key;
+    return TRUE;
+}
+
+// Attempt to hang the current thread if it is a GUI thread, i.e. a thread that
+// has a message queue.
+//
+// If the current thread does not have a message queue, or a re-entrant call is
+// detected, just returns immediately.
+static void hang_gui_thread() {
+    // Use `GetModuleHandleW` and `GetProcAddress` to avoid loading user32.Dell
+    // if it is not already loaded.
+    typedef BOOL (WINAPI *pfIsGUIThread)(BOOL);
+    typedef BOOL (WINAPI *pfGetMessageW)(LPMSG, HWND, UINT, UINT);
+    typedef BOOL (WINAPI *pfTranslateMessage)(const MSG *);
+    typedef LRESULT(WINAPI * pfDispatchMessage)(const MSG*);
+
+    // Guards initialization of `tls_key`.
+    static INIT_ONCE tls_init_once = INIT_ONCE_STATIC_INIT;
+
+    // Thread-local storage key to guard against re-entrant calls to
+    // `_PyThread_hang_thread`, which can occur if a Window procedure called by
+    // `DispatchMessage` attempts to acquire the GIL.
+    static DWORD tls_key;
+
+    MSG msg;
+    HMODULE user32;
+    pfIsGUIThread is_gui_thread;
+    pfGetMessageW get_message;
+    pfTranslateMessage translate_message;
+    pfDispatchMessage dispatch_message;
+    user32 = GetModuleHandleW(L"user32");
+    if (!user32) return;
+    // Query `IsGUIThread` before calling `GetMessageW` to avoid converting this
+    // thread to a GUI thread if it is not already one.
+    is_gui_thread = (pfIsGUIThread)GetProcAddress(user32, "IsGUIThread");
+    if (!is_gui_thread) return;
+    if (!is_gui_thread(FALSE)) return;
+
+    {
+        BOOL should_initialize;
+        PVOID context;
+        if (!InitOnceBeginInitialize(&tls_init_once, /*dwFlags=*/0,
+                                     /*fPending=*/&should_initialize,
+                                     /*lpContext=*/&context)) {
+            // InitOnceBeginInitialize failed for some reason.
+            return;
+        }
+        if (should_initialize) {
+            // Initialize `tls_key`.
+            tls_key = TlsAlloc();
+            InitOnceComplete(&tls_init_once, /*dwFlags=*/0, /*lpCOntext=*/NULL);
+        }
+        if (tls_key == TLS_OUT_OF_INDEXES) {
+            // Failed to initialize `tls_key`.
+            return;
+        }
+    }
+    if (TlsGetValue(tls_key)) {
+        // Re-entrant call detected.
+        return;
+    }
+
+    // Prevent re-entrant calls.
+    TlsSetValue(tls_key, (LPVOID)1);
+
+    get_message = (pfGetMessageW)GetProcAddress(user32, "GetMessageW");
+    if (!get_message) return;
+    translate_message = (pfTranslateMessage)GetProcAddress(user32, "TranslateMessage");
+    if (!translate_message) return;
+    dispatch_message = (pfDispatchMessage)GetProcAddress(user32, "DispatchMessage");
+    if (!dispatch_message) return;
+
+    while (1) {
+        if (GetMessage(&msg, NULL, 0, 0) == -1) return;
+        TranslateMessage(&msg);
+        DispatchMessage(&msg);
```

Comment:
I don't understand why a thread must handle Windows messages. Currently, DispatchMessage() is not called anywhere in Python. I don't see why it should start calling it. What happens if you just run the Sleep(INFINITE) loop?

Reply:
Regarding signals: suppose the application receives a non-blocked signal on a hung thread.  I believe what will happen as I have this pull request currently, is that the signal handler is run as normal.  Is that a problem?

I'm not an expert on Windows, but my first thought was the same as yours, to just call `Sleep(INFINITE);`.  But then I looked at the MSDN documentation for `Sleep`, and it says that calling Sleep can cause deadlock if the thread receives messages.  For that reason I added the message handling.

Reply:
Also, I think Windows is somewhat special compared to other operating systems, in that it has "GUI"-specific things like message handling integrated deeply in the operating system.

Reply:
If a thread owns windows, it needs to execute a message loop to get messages from its message queue and handle/dispatch them. Otherwise a call will hang that sends a message to a window or broadcasts a message to top-level windows, such as `SendMessageW(HWND_BROADCAST, WM_, ...)`. Using `SendMessageTimeoutW()` avoids hanging due to unresponsive threads.

When a thread exits via `PyThread_exit_thread()`, any windows it owns are destroyed. That won't be the case if the code is changed to hang a thread that tries to acquire the GIL at exit. However, I don't know whether it's a problem that's worth addressing. If the process is exiting, the thread will terminate soon. We're talking about hanging a thread for a short time, right?

Reply:
Agreed, this is getting overly complex for a Windows situation we don't even have evidence of being a problem for this particular case.

These code paths are hanging a thread because the main thread has indicated the python interpreter is exiting. Ordinarily that suggests the process is exiting and everything goes away quickly. But in an embedded Python interpreter case it is technically possible for the application to just finalize the interpreter but not exit the process. I don't know how common that is. I expect not common at all. I would not worry about it. At least not for any first iteration on this. The norm for things embedding CPython is to only initialize one and only finalize when the process is finishing. Given that our interpreters are still not friendly to multiple initialize+finalize cycles (we've been improving that a lot, but are still not there yet, so I don't expect any Windows GUI application to depend on doing that).

I think a `Sleep(INFINITE);` is simpler and more appropriate for now.  With a comment suggesting we may need to do Windows even handling stuff _if_ this ever presents itself as a problem for anyone.

Reply:
I have updated the PR to remove the windows message handling.